### ISSUE TITLE
LigerCrossEntropyLoss(reduction='none') cause grad_norm=0

### ISSUE THREAD

**Issue by sanghyuk-choi** [linkedin/Liger-Kernel#678]
### 🐛 Describe the bug

`LigerCrossEntropyLoss(reduction='none')` or `LigerFusedLinearCrossEntropyLoss(reduction='none')` makes grad_norm to zero.

I'm not sure but I think `labels` contains lots of `IGNORE_INDEX`, makes grad_norm to zero.
when `reduction='mean'`, It works fine

### Reproduce

reproduction code
```python
    # override compute_loss from huggingface.Trainer
    # model = LlamaForCausalLM
    def __init__(self, args, **kwargs):
        ....
        self.loss_fct = LigerCrossEntropyLoss(reduction='none')
        ....
    
    @override
    def compute_loss(self, model, inputs, return_outputs=False, **kwargs):       
        loss = None
        labels = inputs.pop("labels") if "labels" in inputs else None
        inputs.update({"return_dict": True})
        outputs = model(**inputs)
        logits = outputs.logits

        if labels is not None:
            # Upcast to float if we need to compute the loss to avoid potential precision issues
            logits = logits.float()
            # Shift so that tokens < n predict n
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            # Flatten the tokens
            shift_logits = shift_logits.view(-1, self.model.config.vocab_size)
            shift_labels = shift_labels.view(-1)
            # Enable model parallelism
            shift_labels = shift_labels.to(shift_logits.device)
            ce_loss = self.loss_fct(shift_logits, shift_labels)
            mask = (shift_labels != -100)
            loss = (ce_loss * mask).sum() / mask.sum()

        return (loss, outputs) if return_outputs else loss
```

### Versions

0.5.5

**Comment by Tcc0403** [comment]
I know `LigerFusedLinearCrossEntropyLoss` doesn't work with `reduction='none'`. (We should take it down)
But I can't reproduce the issue in `LigerCrossEntropyLoss` with `reduction='none'`. Here's my testing:

```python
import torch

from torch.nn import CrossEntropyLoss

from liger_kernel.transformers import LigerCrossEntropyLoss

B, T, V = 2, 2048, 32000

ignore_index = -100
reduction = "none"
device = "cuda"
dtype = torch.float32
scalar = 2

atol, rtol = 1e-8, 1e-5

use_ignore_index = False

if use_ignore_index:
    # Passed
    target_ce = LigerCrossEntropyLoss(ignore_index=ignore_index, reduction=reduction)
    torch_ce = CrossEntropyLoss(ignore_index=ignore_index, reduction=reduction)
else:
    # Also passed
    target_ce = LigerCrossEntropyLoss(reduction=reduction)
    torch_ce = CrossEntropyLoss(reduction=reduction)

_tensor = torch.randn(B * T, V, device=device, dtype=dtype) * scalar
_input = _tensor.detach().clone().requires_grad_(True)
_input2 = _tensor.detach().clone().requires_grad_(True)

target = torch.randint(0, V, (B * T,), device=device, dtype=torch.long)

# Assign some random number of elements as ignore_index
num_elements_to_assign = torch.randint(
    1, B * T // 2, (1,)
).item()  # Random number of elements to set to ignore_index
indices_to_assign = torch.randperm(B * T)[:num_elements_to_assign]  # Randomly select indices
target[indices_to_assign] = ignore_index

if use_ignore_index:
    loss1 = torch_ce(_input, target)
    loss2 = torch_ce(_input2, target)
else:
    output = torch_ce(_input, target)
    output2 = target_ce(_input2, target)
    mask = (target != -100)
    loss1 = (output * mask).sum() / mask.sum()
    loss2 = (output2 * mask).sum() / mask.sum()

assert torch.allclose(loss1, loss2, atol=atol, rtol=rtol)

loss1.backward(gradient=torch.ones_like(loss1))
loss2.backward(gradient=torch.ones_like(loss2))
assert torch.allclose(_input.grad, _input2.grad, atol=atol, rtol=rtol)
```

**Comment by sanghyuk-choi** [comment]
Hi Tcc0403, Thank you for the reply,
I can reproduce by adding a line 

```python
...
target[indices_to_assign] = ignore_index
target[200:600] = ignore_index # a continuous ignore indices label
if use_ignore_index:
...
```

makes an AssertionError.

**Comment by Tcc0403** [comment]
@sanghyuk-choi Thanks for your reply. I'll take a look.

**Comment by wangruohui** [comment]
Hello @Tcc0403 ,

Do you plan to support `reduction='none'` for `LigerFusedLinearCrossEntropyLoss`? I found it useful. 
I saw in #680 it was removed. Is the previous implementation wrong in backward computation?

**Comment by Tcc0403** [comment]
Hi @wangruohui ,

Yes, the previous implementation was wrong. It requires separating gradient calculation from forward pass.
